### PR TITLE
Fix so all existing tests run. Temporarily disable broken call to notary run. Validate all spawned threads finish.

### PR DESCRIPTION
**Pull request by ltfschoen**
* Only 4 out of the 5 existing tests were being picked up when `cargo test` was run previously. Fixed by moving the tests from tests/cli_test.rs into cli/arg.rs as unit tests with their implementation so all 5 of 5 tests run when `cargo test` is run
* Previously whenever `notary.run();` was run in a thread within lib.rs the code would hang and the thread would not finish as no resultant value was observed in the `join` block (due to the loop in notary.rs). I have temporarily commented out the call to `notary.run();` until the code that this calls does not interfere with achieving a final value from its thread so all the unit tests work fully.
* I have used created a couple of vectors that are passed into each thread and iterated over to calculate a result. If all the child threads that are spawned return a final value successfully from `Ok` in their `join` method, then if we sum those values we can test that it adds to the result we expect. I have used this approach and created unit tests that we should expect to always pass if nothing goes wrong in the spawned threads.

### CODE REVIEW

File: src/cli/args.rs [hunk 1]
```diff
@@ -49,3 +49,83 @@ pub fn parse_args(args: Vec<String>) -> Result<config::Config, &'static str> {
...
+        assert_eq!(config_verbose.mode, config::Mode::Both);
+        assert_eq!(config_short.mode, config::Mode::Both);
+        assert_eq!(config_default.mode, config::Mode::Both);
+    }
+
+    #[test]
+    fn it_reports_invalid_arguments() {
+        // Invalid configuration
+        let test_args_configuration = vec![String::from("-bin"), String::from("notary")];
+        let error_configuration = parse_args(test_args_configuration);
```
Comment (jamesray1) [inline src/cli/args.rs]: parse_args is a nightly only experimental API. I think we should use stable Rust in production code.

https://doc.rust-lang.org/nightly/nightly-rustc/syntax_ext/format/fn.parse_args.html

File: src/cli/args.rs [hunk 2]
```diff
@@ -49,3 +49,83 @@ pub fn parse_args(args: Vec<String>) -> Result<config::Config, &'static str> {
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn it_sets_client_mode_to_proposer() {
+        // Verbose command
+        let test_args_verbose = vec![String::from("-mode"), String::from("proposer")];
+        let config_verbose = parse_args(test_args_verbose).unwrap();
```
Comment (jamesray1) [inline src/cli/args.rs]: parse_args is a nightly only experimental API. I think we should use stable Rust in production code.

https://doc.rust-lang.org/nightly/nightly-rustc/syntax_ext/format/fn.parse_args.html

Comment (jamesray1) [reply src/cli/args.rs]: Oh wait parse_args is a function in this file.

**Comment by ltfschoen** [comment]
@ChosunOne has implemented a custom method `parse_args` [here](https://github.com/Drops-of-Diamond/diamond_drops/blob/develop/src/cli/args.rs#L9). @jamesray1 identified that this is also a method name used in nightly https://doc.rust-lang.org/nightly/nightly-rustc/syntax_ext/format/fn.parse_args.html. We would not be using nightly in production anyway, but just for peace of mind I have renamed `parse_args` to `parse_cli_args` to avoid any potential name conflicts

**Comment by ChosunOne** [comment]
The call to notary.run() isn't making the tests hang, its the call to join.  What is happening is both threads are spawning and running, but neither thread currently has an exit condition.  Join blocks the main thread from exiting until the specified thread has completed.  Since currently there is no exit condition for the spawned threads to ever end, the integration test will hang because the main program doesn't exit because the child threads aren't exiting.  

I can add exit conditions so that they do exit for the tests.

**Comment by ltfschoen** [comment]
Sounds good!

**Pull request closed without merging**